### ISSUE TITLE
Nix installer fails on Fedora 28 due to SELinux denial

### ISSUE THREAD

**Issue by e-user** [NixOS/nix#2374]
I have just come across this trying to install Nix on Fedora 28 Workstation with SELinux enabled (default) using the `./install-multi-user` script from the `nix-2.0.4-x86_64-linux` bundle.

```
---- sudo execution ------------------------------------------------------------
I am executing:

    $ sudo systemctl link /nix/var/nix/profiles/default/lib/systemd/system/nix-daemon.service                                                                                

to set up the nix-daemon service

Failed to link unit: Access denied
```

I can confirm this is an SELinux denial due to this AVC message in `/var/log/audit/audit.log`:
```
type=AVC msg=audit(1535267537.912:2983): avc:  denied  { read } for  pid=1 comm="systemd" name="default" dev="nvme0n1p7" ino=2095622 scontext=system_u:system_r:init_t:s0 tcontext=unconfined_u:object_r:default_t:s0 tclass=lnk_file permissive=0 
```

This issue seems to be the same as the one [discussed](https://logs.nix.samueldr.com/nixos-dev/2018-04-12) in #nixos-dev.

Based on the nature of the error, I would naturally assume all SELinux-enabled systems would be affected by this.

**Comment by dezgeg** [comment]
Yeah, probably the installer should bail out if SELinux is enforcing (at least for now).

**Comment by grahamc** [comment]
This bug is especially important now, the default installer as of 2.1 will detect systemd and default to the multi-user installer. See the attached PDF for a demonstration of the problem. (caution: big)
[report-d-2.pdf](https://github.com/NixOS/nix/files/2338464/report-d-2.pdf)

**Comment by grahamc** [comment]
Adding this logic:

```
        if /usr/sbin/sestatus && /usr/sbin/sestatus 2>&1 | grep -q enforcing; then
            echo "punting to no-daemon"
                curl "${installUrl}" | sh -s -- --no-daemon
        else 
                  curl "${installUrl}" | sh
        fi
```

worked in the test harness, but the following did not:

```
        if sestatus && sestatus 2>&1 | grep -q enforcing; then
            echo "punting to no-daemon"
                curl "${installUrl}" | sh -s -- --no-daemon
        else 
                  curl "${installUrl}" | sh
        fi
```

because the install script, when run over SSH, doesn't have `/usr/sbin` in the PATH.

I'm not sure the nicest way to do this, but we could try adding a few directories to the PATH (/sbin, /usr/sbin, others ...?) just for this check.

Any ideas?

**Comment by dezgeg** [comment]
The `getenforce` command should be better for scripting.

Perhaps:
````
getenforce=$(PATH=$PATH:/sbin:/usr/sbin command -v getenforce)
if [ -n "$getenforce" ] && [ "$($getenforce)" = Enforcing ]; then
    echo "No multi-user for you, sorry"
fi
````

**Comment by dezgeg** [comment]
Because I personally don't know anything about writing SELinux rules for Fedora (nor CentOS for that matter.

Don't you also need something to allow systemd to use socket activation for the Nix daemon? At least I ran into similar issue to https://github.com/ganto/copr-lxd/issues/12.

**Comment by e-user** [comment]
Since I need to be able to run Nix based services, I will try my best to provide patches for the installer and possibly nix-store to integrate with SELinux properly.

**Comment by dezgeg** [comment]
Yes, that sounds good. I think we still need some quick solution for Nix 2.1.

**Comment by e-user** [comment]
Newly filed, related issue: https://github.com/systemd/systemd/issues/9997

**Comment by e-user** [comment]
If anyone wants this issue fixed as much as I do, please help me test #2670.

**Comment by sonirico** [comment]
This is still happening, I have the same output when installing nix as the initial comment.

**Comment by teohhanhui** [comment]
Bad @stale bot.

This is obviously still not fixed.

**Comment by yawaramin** [comment]
I feel like I should point out here that the point of the stale bot is not to care about whether or not something is fixed, but to make a guesstimate that something will not be fixed because no one has bothered to fix it till now. To put the issue to rest either way, perhaps a maintainer would be so kind as to close this issue as 'wontfix' and make a note that Nix doesn't support Fedora or other SELinux distros.

**Comment by cgoetzke** [comment]
I think I'm hitting this, too.

Fedora 34 Workstation (KDE Plasma Spin).

Tried a multi-user install with `sh <(curl -L https://nixos.org/nix/install) --daemon` and got (snipped for brevity):

```
---- sudo execution ------------------------------------------------------------
I am executing:

    $ sudo systemctl link /nix/var/nix/profiles/default/lib/systemd/system/nix-daemon.service

to set up the nix-daemon service

Failed to link unit: Access denied

---- oh no! --------------------------------------------------------------------
Jeeze, something went wrong. If you can take all the output and open
an issue, we'd love to fix the problem so nobody else has this issue.

:(

We'd love to help if you need it.

You can open an issue at https://github.com/nixos/nix/issues

Or feel free to contact the team:
 - Matrix: #nix:nixos.org
 - IRC: in #nixos on irc.libera.chat
 - twitter: @nixos_org
 - forum: https://discourse.nixos.org
```

Since this issue has been open for 3.5 years, I'm assuming this isn't likely to get fixed anytime soon.

If it's not going to get a proper fix, it would be nice if the multi-user install would at least check for SELinux at the top and bail before starting to make changes on the system.

I'm also open to suggestions on the best way to back out this partial install I have now.

**Comment by ryleu** [comment]
Can confirm with Fedora 35.

Snippet:

```
---- sudo execution ------------------------------------------------------------
I am executing:

    $ sudo systemctl link /nix/var/nix/profiles/default/lib/systemd/system/nix-daemon.service

to set up the nix-daemon service

Failed to link unit: Access denied

---- oh no! --------------------------------------------------------------------
```
[full output](https://cdn.discordapp.com/attachments/930298475063697448/930301599820161064/message.txt)

My declarative adventure has been put on hold, it seems.

**Comment by emlautarom1** [comment]
Any workaround for Fedora users?

**Comment by EoghanLawless** [comment]
I'm facing the same issue on CentOS Stream 9
```
 $ uname -a
Linux <hostname> 5.14.0-47.el9.x86_64 #1 SMP PREEMPT Sat Jan 22 02:21:23 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux

 $ cat /etc/centos-release
CentOS Stream release 9
```

**Comment by ryleu** [comment]
It seems that this will not be fixed, considering that it’s 3 years old. I’ll just have to find another declarative package manager, I suppose.

**Comment by bjorn** [comment]
So, this just failed for me on Fedora 35 as well. Since the issue has been known for so long, I'm surprised that there isn't  at least an early-out check in the installer, because now it has done a lot of things on my system, which I have no idea how to undo. :-(

Also a warning could be added to the [installation instructions](https://nixos.org/download.html#nix-install-linux) about this problem.

This message from the install script prompted me to go here, but it leaves me a bit speechless finding this old issue:
```
Jeeze, something went wrong. If you can take all the output and open
an issue, we'd love to fix the problem so nobody else has this issue.
```
How many people still need to run into this and are left with a half-finished Nix install on their system?

**Comment by ryleu** [comment]
At least three. I’m incredibly annoyed that it didn’t at least *check* before doing a ton on random stuff to my computer that may leave it in an insecure state, for all I know.

**Comment by kiteloopdesign** [comment]
Same issue here with Fedora 35. Many years later and not even a workaround has been provided. 

I understand they must be swamped with so many github issues, but not properly documenting this is going to cost them so many more issues being submitted

**Comment by matthewpi** [comment]
I was able to get Nix on Fedora Silverblue installed by using these workarounds.

https://gist.github.com/matthewpi/08c3d652e7879e4c4c30bead7021ff73#file-readme-md

**Comment by stelcodes** [comment]
Ran into the same issue. Does anyone know how to manually finish the installation?

**Comment by bjorn** [comment]
> After this, rerun the installation script and it should work.

So you mean, if no cleanup is necessary, the only thing to make the install work is running `sudo setenforce 0` beforehand? Is this change temporary? If not, should or can it be re-enabled after installation? Or would the more specific permission changes by @matthewpi be preferred?

**Comment by ryleu** [comment]
The solution is to create SELinux policies, but the ability to do so is an *exceptionally* rare skill. I am learning how to do so, but it may take me multiple months. I'll send them over once they're ready. From there, the devs can work on getting it merged into distros that use SELinux.

**Comment by adisbladis** [comment]
I have created an alternate set of installers using distribution native packaging formats over https://nix-community.github.io/nix-installers/.
Currently supported distributions are rpm-based, deb-based and pacman-based.
All of these come with selinux policies and it's been tested on Fedora.

I hope that after more testing & polishing this can become our default recommended installation method for the supported distributions.

**Comment by adisbladis** [comment]
> do you intend to add this to Fedora's repository so that users are able to install it via dnf?

No, and I do not think that they would permit a package built in this kind of way and used as a one-shot install rather than a fully distro managed package.

> If the file is intended to remain external, please provide a vanity-link so that I am able to automate the process of acquisition of the installer.

What do you mean by vanity link in this context? A stable URL to the installer or something else?
In any case I've added a `-latest` that works for every installer: https://nix-community.github.io/nix-installers/x86_64/nix-multi-user-latest.deb

**Comment by BEEDELLROKEJULIANLOCKHART** [comment]
@adisbladis, I would nevertheless be incredibly grateful if you were to attempt to add it, not because I particularly desire it, but because recommendation of this is more easy if the distribution has verified that it is not malicious, intentional or not.

The "-latest" is exactly what I desire, but I do not believe that it is specified within "http://nix-community.github.io/nix-installers", so please add it if I am correct.

**Comment by BEEDELLROKEJULIANLOCKHART** [comment]
@adisbladis, the "-latest" is approximately what I desire, but I would prefer a .noarch version so that my code does not require as much verification of the bit-depth of the operating-system of its host.

Additionally, I do not believe that it is specified within "http://nix-community.github.io/nix-installers", so please add that it if I am correct, so that other users learn of this ability.

I am thankful for your assistance.

**Comment by chdoc** [comment]
I just want to note that this it is now 2022 and SELinux on Fedora 35 is so unobtrusive that many people (including me) may not even realize their system is using SELinux. So modifying a host of things all over the system and then bailing at the very end due to an issue that's been known for almost four years is about the gravest sin I have ever seen an installer perform in over a decade of working in IT. 

Once I've cleaned up this mess, I'll probably refrain from installing nix on anything other than throw-away VMs for the foreseeable future. :angry:

**Comment by chdoc** [comment]
Well, let's assume for a moment that I would like to actually get this working the way it should be, that is without switching off SELinux. From the looks of it, the problem is that it tries to "install" `nix-daemon.service` through `systemctl link` (i.e., making the `.service` file available to systemd via a symbolic link). However SELinux forbids systemd from accessing `/nix`. So I suppose there are two options here:
1. add a SELinux rule enabling the `systemctl link` to succeed
2. Install the `.service` file to the default location (`/lib/systemd/system`), this should not require any SELinux rule.  

Option (2) seems preferable to me.

Of course, there may still be an issue with the permissions of the sockets used by the service. But for that I'd need to know more about how that works and what rules (Fedora) SELinux imposes here by default.

**Comment by ahayzen** [comment]
>     2. Install the `.service` file to the default location (`/lib/systemd/system`), this should not require any SELinux rule.

Note that with Fedora Silverblue `/lib` is read-only. And if you disable SELinux it is currently possible to install Nix on Fedora Silverblue with `nix-daemon.service` being linked into `/etc`.

If a `.service` was installed into `/lib` could there be a fallback to install into `/etc` as well if `/lib` isn't writeable ?

**Comment by ahayzen** [comment]
> Do you mean read only in the sense that not even a software install (which anyway requires root privileges) cannot install new files in `/lib` on Fedora Silverblue?

Yes, Fedora Silverblue is an emerging variant of Fedora which uses ostree and is immutable, this means that `/bin /lib /usr` are a read-only filesystem, only `/etc /var` are writeable (`/home` is stored in `/var`). You can see this below

```
$ sudo touch /lib/systemd/system/test
[sudo] password for andrew: 
touch: cannot touch '/lib/systemd/system/test': Read-only file system
```

So if `/lib` is a read-only filesystem it would be good to have a fallback of either linking or writing directly to `/etc` if this becomes the route to install without SELinux denials.

Note that if a rpm of nix was in the fedora repository then Silverblue does allow for "layering" this package into the read-only part of ostree, so in that case `/lib` could be used, but I believe you were describing using the Nix installer.

**Comment by chdoc** [comment]
Thank you for the explanation regarding the Silverblue specifics.

> Note that if a rpm of nix was in the fedora repository then Silverblue does allow for "layering" this package into the read-only part of ostree, so in that case `/lib` could be used, but I believe you were describing using the Nix installer.

I (semi-purposefully) wasn't specific about the mechanism used for the install. However, I think that - from the user perspective -  installing software through the distributions own package manager is always preferable to running an install script. In particular the package manager usually records which installed files belong to which installed package (e.g., by running `rpm -q --whatprovides <file>`) and can therefore simply and cleanly uninstall packages. Of course, I acknowledge that providing (good) installers requires knowledge about both the software to be installed as well as the quirks of the particular distribution (or even release). This is something neither the Nix developers nor the Fedora maintainers may be willing to engage in. :man_shrugging:

**Comment by Artturin** [comment]
https://github.com/NixOS/nix/pull/6639

**Comment by BEEDELLROKEJULIANLOCKHART** [comment]
@outergod should modify the title to demonstrate that this affects Rawhide rather than any specific release.

**Comment by thomasjcf29** [comment]
I've raised #6881 as the solution provided in #6639 prevents install on a system with SELinux set to permissive. For all intents and purposes disabled and permissive should be treated the same (as far as nix is concerned).

I've gone into further detail on the PR.

**Comment by BEEDELLROKEJULIANLOCKHART** [comment]
@thomasjcf29, you might want to link your comments rather than copy them. Makes it easier to maintain them when there's a spelling-mistake, etcetera.

**Comment by pmoieni** [comment]
Is there any fix for this or we should use the single-user install?

**Comment by uxdiin** [comment]
> I have created an alternate set of installers using distribution native packaging formats over https://nix-community.github.io/nix-installers/. Currently supported distributions are rpm-based, deb-based and pacman-based. All of these come with selinux policies and it's been tested on Fedora along with Ubuntu & Arch.
> 
> I hope that after more testing & polishing this can become our default recommended installation method for the supported distributions.
> 
> Please note that this doesn't work for rpm-ostree distros such as Fedora Silverblue, but it works for regular Fedora.

anyone having xdg-desktop-portal error after installing this

**Comment by dnkmmr69420** [comment]
Try out this [guide](https://github.com/dnkmmr69420/nix-with-selinux) it works on silverblue, workstation, and RHEL (tested on rocky linux)

**Comment by cgevans** [comment]
#6639 is useful in preventing failed multi-user installations because of this bug, but when requesting a single-user installation, the installer still (incorrectly) states "Note: a multi-user installation is possible".

**Comment by universish** [comment]
https://github.com/NixOS/nix/issues/11689
https://github.com/NixOS/nixpkgs/issues/350071

**Comment by universish** [comment]
> @chdoc SELinux açıksa, yükleyicinin çok kullanıcılı kurulumdan vazgeçme seçeneğiyle birlikte gerçekten büyük bir uyarı görüntülemesi gerekiyor. Yükleyiciye nispeten basit bir ekleme olmalı, neden eklenmediğinden gerçekten emin değilim. Bunun üzerinde çalışmayı deneyebilirim. Pisliği temizlemek çok zor değil:
> 
> ```
> sudo rm /etc/bashrc
> sudo mv /etc/bashrc.backup-before-nix /etc/bashrc
> sudo rm -r /nix
> ```
> 
> Ardından SELinux'u kapatın: `sudo setenforce 0` Kapalı olduğundan emin olun: `sudo getenforce` Daha sonra Nix'i tekrar yükleyin.
> 
> Elbette artık Nix'le ilgilenmiyorsanız kullanıcıları da temizlemeniz gerekecek.

https://nix.dev/manual/nix/2.18/installation/uninstall

**Comment by universish** [comment]
What is this:
https://nix.dev/manual/nix/2.18/installation/installing-binary#multi-user-installation

**Comment by universish** [comment]
UPPS!!!!

```
macellan@fedora:~$ sudo rm /etc/bashrc
sudo mv /etc/bashrc.backup-before-nix /etc/bashrc
[sudo] password for macellan: 
mv: Failed to state '/etc/bashrc.backup-before-nix': No such file or directory
```
@stelcodes , I deleted /etc/bashrc as per your advice. I tried to restore from the original backup, it says that no such directory or file exists. I wish I had backed up /etc/bashrc before I took your advice. What to do now, can you tell me...

**Comment by bjorn** [comment]
@universish I had a look with `dnf whatprovides /etc/bashrc` and it lists the `setup` package as providing that file. You can re-install that with `sudo dnf reinstall setup` to get back the `/etc/bashrc` file.

**Comment by universish** [comment]
deleting this file broke my terminal. it doesn't say macellan@fedora $. it says bash-5.2$. My commands don't work either. Aren't you ashamed to write this command? @stelcodes

**Comment by universish** [comment]
thanks @bjorn . it's fixed.

**Comment by stelcodes** [comment]
> deleting this file broke my terminal. it doesn't say `macellan@fedora $`. it says `bash-5.2$`. My commands don't work either. Aren't you ashamed to write this command? @stelcodes

No. When I ran into this problem, those are the steps I took to fix it. Apparently the Nix installer did not or could not create the backup `bashrc` on your machine before the installer failed. I edited my comment to only include the `mv` command as that should overwrite `/etc/bachrc` only if the backup exists, avoiding the need to `rm` anything specifically. You'll have to excuse me for not putting up more guardrails into my example command. I had no idea naive sysadmins would blindly type them in and harass me afterwards. Lesson learned!

**Comment by universish** [comment]
Thank you for admitting your mistake, and thank you for having the gall to finally claim that he was harassing you.

**Comment by D3vil0p3r** [comment]
Just come up on this discussion. @outergod is there any news about this open issue?

**Comment by jtrees** [comment]
For those looking for a workaround: [Lix](https://lix.systems/) seems to support SELinux.

**Comment by cole-h** [comment]
(Just FWIW, that isn't unique to Lix -- any recent-enough installer based on https://github.com/DeterminateSystems/nix-installer should inherit our support for SELinux.)

**Comment by devurandom** [comment]
Nix will become available via default Fedora repositories starting from Fedora 44 (currently "rawhide"): https://bugzilla.redhat.com/show_bug.cgi?id=2388768

**Comment by giokhmaladze** [comment]
Nix is available on fedora 41, 42, 43 stable repository already

**Comment by bam80** [comment]
@devurandom @giokhmaladze My concern is the Nix support included in Fedora doesn't help for installing Nix with any non-packaged Nix installers, so in environments where installing rpms is not available (e.g., in distro installer scripts) that "support" is useless.
Correct me if I'm wrong.

**Comment by almereyda** [comment]
Would this also be relevant to using Nix Homemanager, as in defining my Quadlets containers with it, unless

- https://github.com/containers/podman/discussions/20573

is resolved?
